### DISCUSSION TITLE
ScratchAddons and ScratchTools merger?

### DISCUSSION THREAD

**Discussion by Secret-chest** [ScratchAddons/ScratchAddons#5924] [Discussion]
Both extensions are open-source (and they use GPL-3.0) and have many unique useful features. They also seem to have a similar API.

I could help with this, but let's see what ScratchTools thinks.

Steps:

1. Current situation.
2. Remove questionable, outdated, duplicate features from the merge list.
3. Convert remaining ScratchTools to SA format. Let's first use scripts (addon manifest, some APIs etc.) then do it by hand for remaining incompatibilities.
4. Add more addon settings if they make sense (ScratchTools doesn't use settings).
5. Merge. I think only doing this for the "Merge Update" would be enough. 

Should I ask ScratchTools devs if they would like this to happen?

**Comment by cally-jt** [comment]
@rgantzos what do you think?

**Reply by rgantzos** [reply]
I'd love to talk more about this, but at the moment here's how I feel:

- You might as well have both extensions separately. Scratch Addons can use a lot of memory at times, so I don't know how I feel about giving 100+ features to an extension that is already having some trouble (like I said, I'm still open to discussing, and most of this information about Scratch Addons are things that I've heard from its users).
- There's nothing wrong with having 2 extensions. There are already many unique extensions existing for the Scratch website, it doesn't hurt to have a userscript manager such as ScratchTools existing as well.

I just don't see much of a point in merging. The cons outway the pros in my eyes, but I'm open to discussion.

**Reply by Secret-chest** [reply]
ScratchTools only has about 15 unique features.

**Reply by rgantzos** [reply]
> ScratchTools only has about 15 unique features.

Actually it has 111 features... only a few of which are also part of Scratch Addons.

**Reply by WorldLanguages** [reply]
@Secret-chest Not sure what your goal is by creating this discussion and making statements like that...? Not sure if we can get anything constructive out of it.

**Reply by WorldLanguages** [reply]
Just know that if you're interested on adding 15 new features to Scratch Addons, you're welcome to submit issues and PRs.

**Comment by Samq64** [comment]
Scratch Tools became its own extension is partly because at least some of its features were rejected from Scratch Addons. So unless World Languages changes his mind about those, it wouldn't make any sense to merge them.

**Reply by WorldLanguages** [reply]
That might be the main reason ScratchTools was started (can't tell for sure).

But still, that doesn't mean that the APIs need to be different. I think the only specific API that ScratchTools was willing to implement the same way as SA was appendToSharedSpace, but it appears to be unused. I don't think any of their features are using it.

**Reply by Samq64** [reply]
There are probbably other reasons too, otherwise it would just be a rebranded fork of SA.

**Reply by rgantzos** [reply]
I thought I might clarify:
ScratchTools wasn't created because features were being rejected, it was because features were being approved and the developers didn't have the time to meet the work that the community was putting on them to make the new addons, so ScratchTools stepped in. That's just one reason, there are definitely others as Samq64 said.

As for the appendToSharedSpace, that's simply a precaution. Currently, with the limited number of spaces that SA actually uses, there aren't many places where SA and ST could interfere. However, we are migrating towards it.

**Reply by CST1229** [reply]
> ScratchTools wasn't created because features were being rejected, it was because features were being approved and the developers didn't have the time to meet the work that the community was putting on them to make the new addons, so ScratchTools stepped in. That's just one reason, there are definitely others as Samq64 said.

Anyone can contribute to SA, not just the main developers...

**Reply by WorldLanguages** [reply]
Discussion on why ScratchTools was created is probably out of scope here - we should instead try to discuss concrete issues that can have actual impact on developers or extension users.

**Comment by WorldLanguages** [comment]
@Secret-chest Could you clarify your suggestion? The word "merger" might bring confusion.

Are you suggesting that Scratch Addons implements most of ScratchTools' features? There's probably at least 1 of their features that we won't add, so at that point it wouldn't be a "merger" because ScratchTools would continue to have unique features. We wouldn't necessarily need to ask them for permission to do this.

> Should I ask ScratchTools devs if they would like this to happen?

Could you clarify, what would you ask them?

**Comment by Joeclinton1** [comment]
I looked through Scratch Tools set of addons and I believe almost all of the ones which aren't already implemented in ScratchAddons are not worth merging and would cause bloat. I support the merging of:

- Recently Viewed Projects
- NFE Project Checker

**Reply by rgantzos** [reply]
I believe NFE Project Checker was already rejected a few times, but I could be wrong.

**Reply by WorldLanguages** [reply]
I can't say with 100% certainty if we will ever interact with NFE project status.
Providing a way to check recently viewed projects without visiting a studio page could be interesting.
